index.js

```diff
@@ -0,0 +1,282 @@
+            globalModulesSet,
+            sharedChunksSet
+        );
+
+        // If there aren't any modules to add to the chunk don't bother
+        // creating a shared chunk.
+        if (commonModules.size === 0 && moduleDeps.size === 0) {
+            console.warn(
+                `'${spec.name}' chunk not created, contains no modules`
+            );
+            return;
+        }
+
+        if (allChunksNameMap.has(spec.name)) {
+            // Note: this is different from how CommonsChunkPlugin works.
+            throw new Error(
+                `SharedChunksPlugin doesn't work with existing chunks`
+            );
+        }
+
+        // This is the new shared chunk to which we'll be adding modules.
+        const sharedChunk = compilation.addChunk(spec.name);
+
+        // Add module dependencies.
+        for (const dep of moduleDeps) {
+            commonModules.add(dep);
+            globalModulesSet.add(dep);
+        }
+
+        // Add all common modules to the target chunk.
+        for (const module of commonModules) {
+            sharedChunk.addModule(module);
+            module.addChunk(sharedChunk); // modules can appear in multiple chunks
+        }
+
+        // Add chunk dependencies.
+        sharedChunk.parents = [...chunkDeps];
+        for (const chunk of chunkDeps) {
+            chunk.addChunk(sharedChunk);
+        }
+
+        // It's possible that moduleFilter function filters out all modules
+        // from a particular chunk in which case it isn't actually affected.
+        // TODO(kevinb): figure out what to do when extracting common
+        // chunks from async chunks, see CommonsChunkPlugin
+        const actuallyAffectedChunks = new Set();
+        for (const module of commonModules) {
+            for (const chunk of affectedChunks) {
+                if (module.removeChunk(chunk)) {
+                    actuallyAffectedChunks.add(chunk);
+                }
+            }
+        }
+
+        // Copied from makeTargetChunkParentOfAffectedChunks in
+        // CommonsChunkPlugin.js with targetChunk renamed to sharedChunk.
+        for (const chunk of actuallyAffectedChunks) {
+            chunk.parents = [...chunk.parents, sharedChunk];
+            sharedChunk.addChunk(chunk);
+
+            for (const entrypoint of chunk.entrypoints) {
+                entrypoint.insertChunk(sharedChunk, chunk);
+            }
+        }
+
+        sharedChunksSet.add(sharedChunk);
+    }
+
+    // Return module and chunk dependencies for the given commonModules.
+    // Module dependencies are determined recursively.  Whenever a module
+    // appearing in globalModulesSet is encountered, recursion is terminated
+    // and any chunks that the module appears in are return as chunk
+    // dependencies.
+    getDependencies(commonModules, globalModulesSet, sharedChunksSet) {
+        const moduleDeps = new Set();
+        const chunkDeps = new Set();
+
+        const walkDeps = module => {
```

Comment:
I ended up not extracting this as a static global b/c there would be five params which is getting kind of unwieldy.  `getDependencies` only gets called once per shared chunk so defining `walkDeps` here shouldn't be too bad for perf.